eval: partial removal of usages of the stored context

yuzefovich opened the pull request:
This PR contains several commits from #85782 which attempts to remove the stored `context.Context` from `eval.Context` object. I didn't finish that PR since I ran into two very thorny usages, so instead this PR takes some steps towards deprecating the usage of the stored context.

Storing a `context.Context` is an anti-pattern that needs to have a justification for applying, and in case of `eval.Context` I think it was introduced that way specifically for scalar functions evaluations (which should use the same `context.Context`) and that was ok, but over time we introduced many more callsites out of convenience.

Epic: CRDB-20535

Fixes: #46164

yuzefovich committed:
sql: remove some usages of eval.Context.Context

This commit removes some of the explicit usages of
`eval.Context.Context` in favor of explicitly passing the
`context.Context` argument. This is mostly a mechanical change and in
some cases it just pushes the access to `eval.Context.Context` higher up
the stack, but that will make it easier to refactor in the following
commits.

Release note: None

yuzefovich committed:
execinfra: don't use the context via FlowContext

This commit plumbs the context argument to the processor constructors so
that they don't use `flowCtx.EvalCtx.Context`. As an additional minor
benefit the processors now use that context to determine whether they
should collect the execution stats.

Release note: None

yuzefovich committed:
sql: remove the usages of eval.Context.Ctx()

Most of these are in tests.

Release note: None

yuzefovich committed:
eval: propagate context explicitly to GetMultiregionConfig

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89036)
<!-- Reviewable:end -->

yuzefovich committed:
sql: update the eval context with the instrumented context

This commit makes it so that if we derive a new context for the
instrumentation (in order to collect a trace or a statement bundle), we
update the eval context that will be used throughout the execution right
away. This ensures that all builtins will be captured in the trace.

Release note: None

msirek reviewed (approved):
This is great! I used to have doubt about which was the proper `ctx` to use. Explicitly passing it clears up the ambiguity.
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 42 of 42 files at r1, 60 of 60 files at r2, 3 of 3 files at r3, 17 of 17 files at r4, 15 of 15 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89036#-:-NDEFzjr50DmIiamI6XO:b-uxyccp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
Fixed a merge conflict.

bors r+
